src/game/mod.rs

```diff
@@ -269,20 +269,23 @@ impl Game {
 
         //-------------------------------------------------------------------
 
-        let trade_values: Vec<Value> =
+        let mut trade_values: Vec<Option<Value>> =
             (0..player.hand.len())
                 .combinations(2)
                 .filter_map(|pair| player.trade_value(pair[0], pair[1]).ok())
+                .map(|value| value.succ())
                 .collect();
```

Comment:
This can be simplified using the [implementation of `FromIterator<Option<T>>` on `Option<V>` where `V: FromIterator<T>`](https://doc.rust-lang.org/std/iter/trait.FromIterator.html#impl-FromIterator%3COption%3CA%3E%3E):

```rust
let trade_values = /* ... */.collect::<Option<Vec<_>>>();
let mut trade_values = if let Some(values) = trade_values {
    values
} else {
    return false;
};
```
Then you can write the for loop below as:
```rust
if ! trade_values.iter().all(|&value|
    self.table.deck(value).is_empty()
) {
    return false;
}

Reply:
Check out my changes. I tried your code but it didn't work. I wanted the function to return false on `None` and `! deck(value).is_empty()`. The first declaration of trade_values should result in `Vec<Option<Value>>`. Is what you tried to do collect into `Option<Vec<Option<Value>>>`?

Reply:
Explanation: The first `if let` returns `false` on the existence of any `None` in the resulting vector. The rest is self-explanatory.